### PR TITLE
Darcy flow jayjay

### PR DESCRIPTION
**Pull request by jayjay-park**
This is for files under the `datasets` folder.

### CODE REVIEW

File: datasets/configs/darcy1.yaml [hunk 1]
```diff
@@ -0,0 +1,27 @@
+experiment:
+  name: "Darcy Flow Generation"
```
Comment (turquoisedragon2926) [inline datasets/configs/darcy1.yaml]: would be good to group the yamls in a directory. this could get confusing. in the future, dataset yamls of a single generation task are ideally overriding from a base experiment yaml. Also i noticed they all have the same seed? this would lead to same generation on all 4 jobs. hope this was changed during generation.

File: datasets/reduced_orders/fim.py [hunk 2]
```diff
@@ -2,34 +2,117 @@
...
+                    plt.colorbar()
+                    plt.savefig(f'Devito_vjp={j}.png', bbox_inches='tight')
+                    plt.close()
+
+                else:
+                    probe_vector = B[:, j].reshape(y.shape)
+                    Q[:, j] = vjp_func(probe_vector)[0].reshape((simulator.domain,))
 
             U, S, V = torch.linalg.svd(Q)
+            plt.figure(figsize=(5, 5))
```
Comment (turquoisedragon2926) [inline datasets/reduced_orders/fim.py:56]: there shouldnt be plotting logic within this function. the ideal way to plot would be to index into the U post return.

File: datasets/reduced_orders/fim.py [hunk 3]
```diff
@@ -2,34 +2,117 @@
...
                 print(f"Computing FIM Eigenvector {j + 1} of {eigen_count}")
 
-                probe_vector = B[:, j].reshape(y.shape)
-                Q[:, j] = vjp_func(probe_vector)[0].reshape((simulator.domain,))
+                if self.simulator_type == "DARCY":
+                    probe_vector = B[:, j].reshape(x.shape).cpu().numpy()
+                    gradient = simulator.model.compute_gradient(x, probe_vector, p_fwd)
+                    Q[:, j] = torch.tensor(gradient).reshape((simulator.domain,))
+
+                    plt.figure(figsize=(5, 5))
```
Comment (turquoisedragon2926) [inline datasets/reduced_orders/fim.py:44]: same here, no plotting logic, or conditionals

File: datasets/reduced_orders/fim.py [hunk 4]
```diff
@@ -2,34 +2,117 @@
...
-            y, vjp_func = torch.func.vjp(simulator, x)
             eigenvectors = torch.randn((simulator.domain, eigen_count))
+            if self.simulator_type != "DARCY":
+                y, vjp_func = torch.func.vjp(simulator, x)
             
             Z = torch.randn((simulator.range, eigen_count)).to(x.device)
             B, R = torch.linalg.qr(Z)
+
+            if self.simulator_type == "DARCY":
+                x = x.cpu().numpy()
```
Comment (turquoisedragon2926) [inline datasets/reduced_orders/fim.py:30]: this is supposed to be generic implemention. by introducing conditionals, this is now super coupled to only working for the darcy dataset. it would be good to abstract this

File: datasets/reduced_orders/fim.py [hunk 5]
```diff
@@ -2,34 +2,117 @@
...
-    def __init__(self, eigen_value_fraction, eigen_vector_count):
+    def __init__(self, eigen_value_fraction, eigen_vector_count, simulator_type, noise_std, probe_size):
         self.eigen_value_fraction = eigen_value_fraction
         self.eigen_vector_count = eigen_vector_count
+        self.probe_size = probe_size   # number of columns m per probe
         self.thread_lock = threading.Lock()
+        self.simulator_type = simulator_type
+        self.noise_std = noise_std
 
     def get_direction(self, simulator, x):
```
Comment (turquoisedragon2926) [inline datasets/reduced_orders/fim.py:19]: actually, looking at `generate_dataset` is this class even used anymore? I dont see it being used

File: train/inversion.py [hunk 6]
```diff
@@ -0,0 +1,720 @@
...
+import torch.nn.functional as F
+from groundwater.devito_op import GroundwaterModel, GroundwaterLayer
+import h5py
+import pickle
+import os
+import random
+
+from models.ns_inversion import NSModel  # Your model
+from utils import get_dataset, load_config, get_model  # Your utils
+
```
Comment (turquoisedragon2926) [inline train/inversion.py:23]: all the inversion seem to have a lot of shared code, it would be good to have utilities that all the scripts rely on.
Inversion should generally be launched from the config, but scripts are fine too. it would be better to have them all in a sepearte directory called `evaluation` or `inversion` at the root level.

File: datasets/utils.py [hunk 7]
```diff
@@ -36,69 +37,156 @@ def create_simulator(model_type, simulator_settings):
...
+    Q_list = []
+    # Sigma = reduced_model.build_anisotropic_covariance(d, sigma_2=1.0, lx=0.2, ly=0.05, device='cpu')
+    # coords = reduced_model.build_coords(d)
+    # Sigma = matern_kernel(coords, nu=1.5, sigma2=1.0, lx=8.0, ly=4.0)
+
+
+    for b in range(num_subsamples):
+        print("subsample", b+1, "of", num_subsamples)
+        x_b = simulator.sample().detach().cpu().numpy()
+        Qb = reduced_model.compute_score_matrix(simulator, x_b, L)  # [p, r]
```
Comment (turquoisedragon2926) [inline datasets/utils.py]: i see, only the score matrix is used. yea, this is not the right place for this logic. this is meant to be a utility file`utils.py`, so this makes the codebase quite coupled and far from modular. The ideal way to do it would be to have a sepearte class which is a funciton of simulator and probing configuation. for future ref
